src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs

```diff
@@ -84,40 +48,57 @@ private static bool TryGetStringValue(string switchName, string envVariable, [No
         {
-            int length = baseName.Length + suffix.Length + version.Length + extension.Length;
-
-            // We validate that suffix and version are not larger than 53 characters.
-            Span<char> result = stackalloc char[length];
-            baseName.CopyTo(result);
-
-            Span<char> secondPart = result.Slice(baseName.Length);
-            suffix.CopyTo(secondPart);
+            ReadOnlySpan<char> icuSuffix = default;
+            ReadOnlySpan<char> version = default;
 
-            Span<char> middle = secondPart.Slice(suffix.Length);
-
-            if (!versionAtEnd)
+            // Custom built ICU can have a suffix on the name, i.e: libicuucmyapp.so.67.1
+            // So users would set the runtime switch as: myapp:67.1
+            int indexOfSeparator = icuSuffixAndVersion.IndexOf(':');
+            if (indexOfSeparator >= 0)
             {
-                version.CopyTo(middle);
+                icuSuffix = icuSuffixAndVersion.AsSpan().Slice(0, indexOfSeparator);
 
-                Span<char> end = middle.Slice(version.Length);
-                extension.CopyTo(end);
+                if (icuSuffix.Length > 35)
+                {
+                    Environment.FailFast($"The resolved \"{icuSuffix.ToString()}\" suffix from System.Globalization.AppLocalIcu switch has to be < 20 chars long.");
+                }
+
+                if (icuSuffix.Length + 1 <= icuSuffixAndVersion.Length)
+                    version = icuSuffixAndVersion.AsSpan().Slice(icuSuffix.Length + 1);
             }
             else
             {
-                extension.CopyTo(middle);
+                version = icuSuffixAndVersion;
+            }
+
+            if (version.Length > 33)
+            {
+                Environment.FailFast($"The resolved version \"{version.ToString()}\" from System.Globalization.AppLocalIcu switch has to be < 33 chars long.");
+            }
 
-                Span<char> end = middle.Slice(extension.Length);
-                version.CopyTo(end);
+            if (suffixWithSeparator)
+            {
+                int suffixLength = icuSuffix.Length + 1;
+                Span<char> finalSuffix = stackalloc char[suffixLength];
+                icuSuffix.CopyTo(finalSuffix);
+                finalSuffix[suffixLength - 1] = '.';
+                LoadAppLocalIcuCore(version, finalSuffix);
+                return;
             }
 
-            return result.ToString();
+            LoadAppLocalIcuCore(version, icuSuffix);
         }
 
-        private static IntPtr LoadLibrary(string library, Assembly assembly, bool failOnLoadFailure)
+        private static string CreateLibraryName(ReadOnlySpan<char> baseName, ReadOnlySpan<char> suffix, ReadOnlySpan<char> extension, ReadOnlySpan<char> version, bool versionAtEnd = false) =>
+            versionAtEnd ?
+                string.Concat(baseName, suffix, extension, version) :
```

Comment:
nit: could $"{baseName}{suffix}{extension}{version}" be better?

Reply:
Spans do not work with interpolated strings today.